### ISSUE TITLE
Expose a real_t type definition in C#

### ISSUE THREAD

**Issue by DeeJayLSP** [godotengine/godot-proposals#8628]
### Describe the project you are working on

A game/plugin that makes use of functions involving Godot's `real_t` in C#

### Describe the problem or limitation you are having in your project

`real_t` is a type defined in the source code for floating point precision. If the engine is built with `precision=double`, then `real_t` becomes a double precision floating point type, otherwise it becomes single precision. This doesn't affect coding in GDScript but is mapped to `double` and `float` types in C#.

In C#, the `float` and `double` types can't be implicitly converted to each other, so in case we're developing a plugin where a built-in function receives the `real_t` type, using the intended type for that build makes it incompatible with the other:

```cs
Vector2 PositionF = new(1.0, 1.0); // Error without cast on single, works on double;
Vector2 PositionD = new(1.0f, 1.0f); // Error without cast on double, works on single;
```

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Simply expose a `real_t` type definition that maps itself to the intended type.

This way, plugin developers may simply cast floating point values to `real_t`, and it gets mapped automatically to the intended type.

Note that it somewhat conflicts with godotengine/godot#74070, but maybe is a solution that doesn't break compatibility.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Right now, the following lines of code (that are already included in [some C# demos](https://github.com/godotengine/godot-demo-projects/blob/4.0/mono/2.5d/addons/node25d-cs/Basis25D.cs) but without `global`) can be introduced for it to work:
```cs
#if GODOT_REAL_T_IS_DOUBLE
global using real_t = System.Double;
#else
global using real_t = System.Single;
#endif
```

So...
```cs
Vector2 PositionR = new((real_t)1.0, (real_t)1.0); // Works on any precision
```

While not that convenient, this allows code to be compatible with both build precisions.

[This is somewhat already in the engine](https://github.com/godotengine/godot/blob/4.2-stable/modules/mono/glue/GodotSharp/GodotSharp/GlobalUsings.cs), so can't it just be exposed as a type definition?

### If this enhancement will not be used often, can it be worked around with a few lines of script?

The few lines of code are already above, but it would have to be copypasted in every project that intends to use it.

### Is there a reason why this should be core and not an add-on in the asset library?

Because this is somewhat already in core.

**Comment by paulloz** [comment]
My two cents. I might be wrong on any or all of those.

1. Parity: do other scripting languages have direct access to `real_t`? I'm not sure about GDExtensions, but I don't believe that's a thing in GDScript.
2. Exposing type aliases outside a project is not straightforward: aliases are local to their defining files, global aliases are local to their defining project. Outside of those they are unaliased to the underlying type. One of the only ways I can imagine right now would be to leverage a source generator to write a global alias in the game project. And that feels real wrong to me.
3. Use cases in real world: is the need to be precision-agnostic that common? IMHO most game projects will have their codebase intended to work with either a version of Godot built with `single`, or `double`, but won't need to be agnostic. I can see the argument when making a library/middleware. But at this point can't it be the responsibility of that library to use `GODOT_REAL_T_IS_DOUBLE` for their types internally?
4. In any case, it should definitely be documented.

**Comment by paulloz** [comment]
> it is exposed in godot-cpp, and the precision type is also available for GDExtension

Nice to know 🙏 Thank you.

Additional note, even with some kind of `real_t` exposed, you'd have to reference a different version of GodotSharp depending which precision you're using (and we only distribute it with single precision on NuGet at the moment).

**Comment by AThousandShips** [comment]
> Additional note, even with some kind of real_t exposed, you'd have to reference a different version of GodotSharp depending which precision you're using 

That's true already the precision version has to match for things to work, no? Can you actually use a different precision build than the engine?

**Comment by AThousandShips** [comment]
Yes, so declaring `real_t` doesn't change anything, you still have to match the build flags, so this requirement is already there

**Comment by Shadowblitz16** [comment]
Please no, the godot collections and variant system in C# was a mistake.
You should making godot work better with C# not limit C# to godot.
